src/Data/Parameterized/Fin.hs

```diff
@@ -0,0 +1,84 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE EmptyCase #-}
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE TypeOperators #-}
+
+{-|
+Copyright        : (c) Galois, Inc 2014-2019
+
+Finite types.
```

Comment:
Could you provide more of an explanation and/or some examples here?  This becomes a sufficiently abstract representation that the utility is not immediately obvious.  It might be nice to add something to the README for p-u as well describing this.

Also, I think the copyright date on this particular module should be "2021".

Reply:
Yes, will do! In short, my motivation for introducing these is to use them as indices into `Vector`.

Reply:
I added more detail here as well as a blurb in the README, let me know what you think.

Reply:
That's helpful, thanks!